### PR TITLE
Fix Unchecked Arithmetic in Diem-Core

### PR DESCRIPTION
**Pull request by terrytzh**
## Motivation

Making money move more freely for more people!

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Related PRs

None

## If targeting a release branch, please fill the below out as well

Fix unchecked arithmetic to prevent overflow.

### CODE REVIEW

File: secure/storage/vault/src/lib.rs [hunk 1]
```diff
@@ -369,7 +371,13 @@ impl Client {
 
         // Trim keys if too many versions exist
         if (max_version - min_version) >= MAX_NUM_KEY_VERSIONS {
-            let min_available_version = max_version - MAX_NUM_KEY_VERSIONS + 1;
+            // let min_available_version = max_version - MAX_NUM_KEY_VERSIONS + 1;
+            let min_available_version = max_version
+                .checked_sub(MAX_NUM_KEY_VERSIONS)
+                .and_then(|n| n.checked_add(1))
+                .ok_or_else(|| {
+                    Error::OverflowError(String::from("trim_key_versions::min_available_version"))
```
Comment (JoshLind) [inline secure/storage/vault/src/lib.rs:379]: nit: you could also use `into` to do the string conversion for you. It's shorter: https://doc.rust-lang.org/std/convert/trait.Into.html, e.g.,:

`Error::OverflowError("trim_key_versions::min_available_version".into())`

File: secure/net/src/lib.rs [hunk 2]
```diff
@@ -499,12 +501,14 @@ impl NetworkStream {
...
-        let mut total_written = 0;
+        let mut total_written: u8 = 0;
 
         while !unwritten.is_empty() {
             let written = self.stream.write(unwritten)?;
-            total_written += written;
-            unwritten = &data[total_written..];
+            total_written = total_written
+                .checked_add(written as u8)
+                .ok_or_else(|| Error::OverflowError(String::from("write_all::total_written")))?;
```
Comment (JoshLind) [inline secure/net/src/lib.rs:510]: nit: you could use `into`. See my comment below.

File: secure/storage/vault/src/lib.rs [hunk 3]
```diff
@@ -50,6 +50,8 @@ pub enum Error {
     SerializationError(String),
     #[error("Synthetic error returned: {0}")]
     SyntheticError(String),
+    #[error("Overflow error: {0}")]
+    OverflowError(String),
```
Comment (JoshLind) [inline secure/storage/vault/src/lib.rs:54]: We normally try to keep these in alphabetical order. Could you move it up?

File: secure/net/src/lib.rs [hunk 4]
```diff
@@ -144,6 +144,8 @@ pub enum Error {
     NoActiveStream,
     #[error("Remote stream cleanly closed")]
     RemoteStreamClosed,
+    #[error("Overflow error: {0}")]
+    OverflowError(String),
```
Comment (JoshLind) [inline secure/net/src/lib.rs:148]: We normally try to keep these in alphabetical order. Could you move it up?

File: common/short-hex-str/src/lib.rs [hunk 5]
```diff
@@ -126,23 +140,25 @@ mod test {
     use std::{str, u8};
 
     #[test]
-    fn test_hex_encode() {
+    fn test_hex_encode() -> Result<(), Error> {
```
Comment (JoshLind) [inline common/short-hex-str/src/lib.rs:143]: For tests, you don't need to return the result. Instead, we often just use `unwrap` or `expect` to ensure that an error is not thrown. If one is, the test will panic and fail. For example you could just do:

`hex_encode(&src, &mut actual).unwrap()`

It also keeps the tests cleaner/simpler.

File: common/short-hex-str/src/lib.rs [hunk 6]
```diff
@@ -81,22 +85,32 @@ const HEX_CHARS_LOWER: &[u8; 16] = b"0123456789abcdef";
 /// Format a byte as hex. Returns a tuple containing the first character and then
 /// the second character as ASCII bytes.
 #[inline(always)]
-fn byte2hex(byte: u8) -> (u8, u8) {
-    let hi = HEX_CHARS_LOWER[((byte >> 4) & 0x0f) as usize];
+fn byte2hex(byte: u8) -> Result<(u8, u8), Error> {
+    let hi = HEX_CHARS_LOWER[(byte
+        .checked_shr(4)
+        .ok_or_else(|| Error::OverflowError(String::from("byte2hex::byte")))?
```
Comment (JoshLind) [inline common/short-hex-str/src/lib.rs:91]: nit: you could use `into`. See my comment below.

File: common/short-hex-str/src/lib.rs [hunk 7]
```diff
@@ -81,22 +85,32 @@ const HEX_CHARS_LOWER: &[u8; 16] = b"0123456789abcdef";
...
 #[inline(always)]
-fn hex_encode(src: &[u8], dst: &mut [u8]) {
-    debug_checked_precondition!(dst.len() == 2 * src.len());
+fn hex_encode(src: &[u8], dst: &mut [u8]) -> Result<(), Error> {
+    debug_checked_precondition!(
+        dst.len()
+            == src
+                .len()
+                .checked_mul(2)
+                .ok_or_else(|| Error::OverflowError(String::from("hex_encode::src")))?
```
Comment (JoshLind) [inline common/short-hex-str/src/lib.rs:105]: nit: you could use `into`. See my comment below.

File: common/short-hex-str/src/lib.rs [hunk 8]
```diff
@@ -81,22 +85,32 @@ const HEX_CHARS_LOWER: &[u8; 16] = b"0123456789abcdef";
...
-    (hi, lo)
+    Ok((hi, lo))
 }
 
 /// Hex encode a byte slice into the destination byte slice.
 #[inline(always)]
-fn hex_encode(src: &[u8], dst: &mut [u8]) {
-    debug_checked_precondition!(dst.len() == 2 * src.len());
+fn hex_encode(src: &[u8], dst: &mut [u8]) -> Result<(), Error> {
+    debug_checked_precondition!(
```
Comment (JoshLind) [inline common/short-hex-str/src/lib.rs:100]: The formatting/line-wrapping here looks interesting.. Have you run the auto formatter?

Something like `cargo x check` should work (it's an internal tool that wraps the rust auto formatters). I'll also run canary on the PR to see if the tests pass with these changes 😄

**Review by JoshLind — CHANGES_REQUESTED** [review]
Thanks, Terry! Looks great. 😄

A couple comments and nits to address, but nothing major. Let's also run the canary process which should run all the land blocking tests and format checks to make sure everything passes.

/canary

**Comment by terrytzh** [comment]
Hi Joshua, just fixed the above mentioned issues and changed the function signatures in short-hex-str back to original as well as switched to using expect in error handling in short-hex-str specifically (don't wanna break codes that use this lib!). Would you help me check if any additional changes are needed & rerun the tests in canary?

File: common/short-hex-str/src/lib.rs [hunk 9]
```diff
@@ -16,8 +16,10 @@ use thiserror::Error;
 pub struct ShortHexStr([u8; ShortHexStr::LENGTH]);
 
 #[derive(Error, Debug)]
-#[error("Input bytes are too short")]
-pub struct InputTooShortError;
+pub enum Error {
+    #[error("Input bytes are too short")]
+    InputTooShortError,
+}
```
Comment (bmwill) [inline common/short-hex-str/src/lib.rs:22]: Can you revert this change since you no longer have another error variant?

Comment (terrytzh) [reply common/short-hex-str/src/lib.rs:22]: Sure, reverting it rn

File: common/short-hex-str/src/lib.rs [hunk 10]
```diff
@@ -82,15 +82,24 @@ const HEX_CHARS_LOWER: &[u8; 16] = b"0123456789abcdef";
 /// the second character as ASCII bytes.
 #[inline(always)]
 fn byte2hex(byte: u8) -> (u8, u8) {
-    let hi = HEX_CHARS_LOWER[((byte >> 4) & 0x0f) as usize];
+    let hi = HEX_CHARS_LOWER[(byte
+        .checked_shr(4)
+        .expect("Integer Overflow at byte2hex")
```
Comment (JoshLind) [inline common/short-hex-str/src/lib.rs:87]: Aah.. I think you misunderstood my last comment here. We indeed want `byte2hex` to return a `Result<(u8, u8), Error>` as you had it before, but in the unit tests, we don't need to return the result when we call this function. We simply just want to call `unwrap()` or `expect()` in the unit test. Does that make sense?

The problem with calling the `expect()` here is that non-test code that relies on this function will panic, which isn't great. So, all you need to do is bring back the old logic here and then for the unit tests that call this function, `unwrap()` there. Does that make sense?

The same goes for `hex_encode` below

Comment (JoshLind) [reply common/short-hex-str/src/lib.rs:87]: Once you've done that you can call `/canary` on this PR and see how it goes!

Comment (terrytzh) [reply common/short-hex-str/src/lib.rs:87]: The two functions originally return u8 tuple and nothing, and the previous changes serve as a temporary error handling measure to line up with error handling in other lib files; changing their return type to Result type at this point might cause all codes that call these two functions to break. I'll commit the other two files for now

**Comment by JoshLind** [comment]
LGTM! Let's canary to see if there's any issues with tests/Libya, and if not, I'll approve!

File: common/short-hex-str/src/lib.rs [hunk 11]
```diff
@@ -16,8 +16,12 @@ use thiserror::Error;
 pub struct ShortHexStr([u8; ShortHexStr::LENGTH]);
 
 #[derive(Error, Debug)]
-#[error("Input bytes are too short")]
-pub struct InputTooShortError;
+pub enum Error {
+    #[error("Input bytes are too short")]
+    InputTooShortError(),
```
Comment (JoshLind) [inline common/short-hex-str/src/lib.rs:21]: nit: this doesn't need empty brackets (as it takes no arguments), just defining the type should be enough, e.g.:

`InputTooShortError,`

File: common/short-hex-str/src/lib.rs [hunk 12]
```diff
@@ -27,18 +31,18 @@ impl ShortHexStr {
     ///
     /// Returns `Err(InputTooShortError)` if the input byte slice length is less
```
Comment (JoshLind) [inline common/short-hex-str/src/lib.rs:32]: nit: could you update the comment to say "Returns Err(Error::InputTooShortError)"

File: common/short-hex-str/src/lib.rs [hunk 13]
```diff
@@ -27,18 +31,18 @@ impl ShortHexStr {
...
 
             // We include a tiny hex encode here instead of using the `hex` crate's
             // `encode_to_slice`, since the compiler seems unable to inline across
             // the crate boundary.
-            hex_encode(&src_short_bytes, &mut dest_bytes);
+            hex_encode(&src_short_bytes, &mut dest_bytes)?;
             Ok(Self(dest_bytes))
         } else {
-            Err(InputTooShortError)
+            Err(Error::InputTooShortError())
```
Comment (JoshLind) [inline common/short-hex-str/src/lib.rs:45]: nit: this should become `Err(Error::InputTooShortError)`

File: common/short-hex-str/src/lib.rs [hunk 14]
```diff
@@ -81,22 +85,32 @@ const HEX_CHARS_LOWER: &[u8; 16] = b"0123456789abcdef";
...
-    (hi, lo)
+    Ok((hi, lo))
 }
 
 /// Hex encode a byte slice into the destination byte slice.
 #[inline(always)]
-fn hex_encode(src: &[u8], dst: &mut [u8]) {
-    debug_checked_precondition!(dst.len() == 2 * src.len());
+fn hex_encode(src: &[u8], dst: &mut [u8]) -> Result<(), Error> {
+    debug_checked_precondition!(
```
Comment (JoshLind) [inline common/short-hex-str/src/lib.rs:100]: nit: you could tidy this up by creating a temporary variable, e.g.:

```
    let expected_length = src
        .len()
        .checked_mul(2)
        .ok_or_else(|| Error::OverflowError("hex_encode::src".into()))?;
    debug_checked_precondition!(dst.len() == expected_length);
```

File: secure/net/src/lib.rs [hunk 15]
```diff
@@ -499,12 +501,14 @@ impl NetworkStream {
     /// This wraps around that buffer and blocks until all the data has been pushed.
     fn write_all(&mut self, data: &[u8]) -> Result<(), Error> {
         let mut unwritten = data;
-        let mut total_written = 0;
+        let mut total_written: u8 = 0;
```
Comment (JoshLind) [inline secure/net/src/lib.rs:504]: I think the problem you're facing is that `u8` isn't large enough. This function is used to write data on a network stream, so it's possible you're hitting `u8::MAX` and this function is failing. Change this to `u64` -- I think that should fix the error here.

**Review by JoshLind — COMMENTED** [review]
A few comments -- I think the failures you're seeing might be in relation to a genuine overflow as a result of the fact you limited the number of written bytes to `u8`. See my comments below 😄 

Also, let's merge all your commits into a single commit to keep a cleaner history. All of the commits relate to the same unit of work.

**Review by JoshLind — APPROVED** [review]
Thanks @terrytzh! Looks good! 😄 

I'll approve and land! 🚢

File: common/short-hex-str/src/lib.rs [hunk 16]
```diff
@@ -81,22 +85,30 @@ const HEX_CHARS_LOWER: &[u8; 16] = b"0123456789abcdef";
 /// Format a byte as hex. Returns a tuple containing the first character and then
 /// the second character as ASCII bytes.
 #[inline(always)]
-fn byte2hex(byte: u8) -> (u8, u8) {
-    let hi = HEX_CHARS_LOWER[((byte >> 4) & 0x0f) as usize];
+fn byte2hex(byte: u8) -> Result<(u8, u8), Error> {
+    let hi = HEX_CHARS_LOWER[(byte
+        .checked_shr(4)
```
Comment (JoshLind) [inline common/short-hex-str/src/lib.rs:90]: @terrytzh, for what it's worth, @bmwill pointed out that this overflow shouldn't ever be possible given the nature of `u8` and shift right. 

However, until we have some way of notifying the linter/arithmetic checker that we've manually reviewed this overflow and it's okay as-is (I'm not aware of a fine-grained way of notifying the arithmetic checker), let's leave it as is.

**Review by JoshLind — COMMENTED** [review]
/land

File: common/short-hex-str/src/lib.rs (hunk 16 above)
Comment (JoshLind) [reply common/short-hex-str/src/lib.rs:90]: Thanks @phlip9!

> surprisingly, the compiler is actually able to (almost) completely remove both the checked arithmetic and error paths (since they're impossible) and my local micro benches only have a ~12% regression here.

What regression are you referring to here? Compiler non-optimizations vs. optimizations, or something else?

Thanks for following up with the revert: https://github.com/diem/diem/pull/8509. Will trust you that these paths are safe 😄

Comment (phlip9) [reply common/short-hex-str/src/lib.rs:90]: I have an old set of micro benchmarks sitting around for when I wrote short-hex-str. they were never published (they're a bit messy). I just added a new case with the error handling and checked arithmetic and compared the benches with optimizations etc. 😅
